### PR TITLE
4.x: Improve AutoDetachObserver memory footprint

### PR DESCRIPTION
**Pull request by akarnokd**
This PR inlines the `SingleAssignmentDisposable` behavior into the `AutoDetachObserver` to save on an allocation and the indirection.

### CODE REVIEW

File: Rx.NET/Source/src/System.Reactive/Internal/AutoDetachObserver.cs [hunk 1]
```diff
@@ -18,7 +20,17 @@ public AutoDetachObserver(IObserver<T> observer)
 
         public IDisposable Disposable
         {
-            set { _disposable.Disposable = value; }
+            set
+            {
+                if (Interlocked.CompareExchange(ref disposable, value, null) != null)
```
Comment (danielcweber) [inline Rx.NET/Source/src/System.Reactive/Internal/AutoDetachObserver.cs]: I wonder whether we can refactor that reoccuring logic into extensions that take refs, taking the idea from [AssignmentDisposable](https://github.com/danielcweber/Rx.NET/blob/3ed55448d0d80da5833673cea19ed9ab68c0cfe3/Rx.NET/Source/src/System.Reactive/Disposables/AssignmentDisposable.cs) a bit further.

Comment (danielcweber) [reply Rx.NET/Source/src/System.Reactive/Internal/AutoDetachObserver.cs]: The Setter-extensions could be called "SetSingle", "SetMultiple" etc....

Comment (akarnokd) [reply Rx.NET/Source/src/System.Reactive/Internal/AutoDetachObserver.cs]: Yes, I do this in my [repo](https://github.com/akarnokd/reactive-extensions/blob/master/reactive-extensions/tools/DisposableHelper.cs) but this is dotnet/reactive. 

The problem with such PRs is that if you need an utility for multiple other PRs as well, either it has to be PRd separately and then merged before anything in new PRs can use it.

Comment (danielcweber) [reply Rx.NET/Source/src/System.Reactive/Internal/AutoDetachObserver.cs]: >Yes, I do this in my repo but this is dotnet/reactive.

Awesome!

>...then merged before anything in new PRs can use it.

Let's hope any of our work will be merged eventually. Until then, we could agree on some basic improvements (like the import of ValueTuples, those extensions etc.) on a separate branch and rebase all our PRs on that branch.

**Pull request merged**